@here/harp-text-canvas/lib/typesetting/SmoothPathTypesetter.ts

```diff
@@ -0,0 +1,323 @@
+            const a2 = this.m_tempAngleDelta[i];
+            const a3 = this.m_tempAngleDelta[i + 1 > e ? 2 * e + 1 - (i + 1) : i + 1];
+            const a4 = this.m_tempAngleDelta[i + 2 > e ? 2 * e + 1 - (i + 2) : i + 2];
+
+            const angleDeltaAvg = (a0 + a1 + a2 + a3 + a4) / 5;
+
+            currentAngle += angleDeltaAvg;
+
+            normal.set(-Math.sin(currentAngle), Math.cos(currentAngle));
+
+            const pb = textPoint;
+            const pa = this.m_tempPoints[i - 1] || pb;
+            const pc = this.m_tempPoints[i + 1] || pb;
+
+            v1.subVectors(pb, pa);
+            v2.subVectors(pc, pa);
+
+            const offset = v2
+                .multiplyScalar(v2.dot(v1) / v2.dot(v2))
+                .add(pa)
+                .sub(pb)
+                .dot(normal);
+
+            normal.multiplyScalar(normalDisplacement + offset);
+            this.m_tempPathPosition.set(normal.x + textPoint.x, normal.y + textPoint.y, position.z);
+            textRenderStyle.rotation = defaultGlyphRotation + currentAngle;
+
+            // Compute various rendering parameters for this glyph.
+            const glyphFont = glyphData.font;
+            const glyphFontMetrics = glyphFont.metrics;
+            const fontStyle = textRenderStyle.fontStyle;
+
+            const isBoldEmulated =
+                (fontStyle === FontStyle.Bold && glyphFont.bold === undefined) ||
+                (fontStyle === FontStyle.BoldItalic &&
+                    glyphFont.bold === undefined &&
+                    glyphFont.boldItalic === undefined);
+            const isItalicEmulated =
+                (fontStyle === FontStyle.Italic && glyphFont.italic === undefined) ||
+                (fontStyle === FontStyle.BoldItalic &&
+                    glyphFont.italic === undefined &&
+                    glyphFont.boldItalic === undefined);
+
+            const isSmallCaps = this.m_tempSmallCaps
+                ? smallCapsArray![i] && textRenderStyle.fontVariant === FontVariant.SmallCaps
+                : false;
+            const smallCapsScale = isSmallCaps
+                ? glyphFontMetrics.xHeight / glyphFontMetrics.capHeight
+                : 1.0;
+            const glyphScale = this.m_tempScale * smallCapsScale;
+
+            const emulationWeight =
+                ((isBoldEmulated ? 0.02 : 0.0) + (isSmallCaps ? 0.01 : 0.0)) *
+                (fontCatalog.size / fontCatalog.distanceRange);
+            const bgWeight =
+                (0.5 * this.m_tempPixelBgSize!) /
+                (fontCatalog.distanceRange * Math.max(glyphScale, 1.0));
+
+            const isMirrored = this.m_tempMirrored[i];
+
+            const verticalOffset =
+                glyphFontMetrics.lineHeight -
+                glyphFontMetrics.base -
+                glyphFontMetrics.distanceRange * 0.5;
+
+            // Compute the glyphs transformation matrix and apply to all corners of a glyph.
+            TypesettingUtils.computeGlyphTransform(
```

Comment:
I really don't like the fact that we have to transform the glyphs twice, specially when this technique only have visible effects around the vertices of the path and not the edges (since the angle delta stays the same inside an edge)...

Cannot we try to pre-compute some information regarding where the path vertices are found in the path's `[0...1]` range and then use this info to modify the glyph placement (in `placeGlyph`) in a similar way ONLY if we fall in the area of effect of any of the vertices?

Think of it as if we were implementing FXAA in the path domain: https://en.wikipedia.org/wiki/Fast_approximate_anti-aliasing#Processes

Reply:
We are doing this once.